adapters/repos/db/lsmkv/snapshot.go

```diff
@@ -0,0 +1,97 @@
+package lsmkv
+
+import (
+	"context"
+	"io/fs"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+// PauseCompaction waits for all ongoing compactions to finish,
+// then makes sure that no new compaction can be started.
+//
+// This is a preparatory stage for taking snapshots.
+//
+// A timeout should be specified for the input context as some
+// compactions are long-running, in which case it may be better
+// to fail the backup attempt and retry later, than to block
+// indefinitely.
+func (b *Bucket) PauseCompaction(ctx context.Context) error {
+	compactionHalted := make(chan struct{})
+
+	go func() {
+		b.disk.stopCompactionCycle <- struct{}{}
+		compactionHalted <- struct{}{}
+	}()
+
+	select {
+	case <-ctx.Done():
+		// resume the compaction cycle, as the
+		// context deadline was exceeded
+		defer b.disk.initCompactionCycle(DefaultCompactionInterval)
+		return errors.Wrap(ctx.Err(), "long-running compaction in progress")
+	case <-compactionHalted:
+		return nil
+	}
+}
+
+// FlushMemtable flushes any active memtable and returns only once the memtable
+// has been fully flushed and a stable state on disk has been reached.
+//
+// This is a preparatory stage for taking snapshots.
+//
+// A timeout should be specified for the input context as some
+// flushes are long-running, in which case it may be better
+// to fail the backup attempt and retry later, than to block
+// indefinitely.
+func (b *Bucket) FlushMemtable(ctx context.Context) error {
+	defer b.initFlushCycle()
+	flushed := make(chan struct{})
+
+	go func() {
+		b.stopFlushCycle <- struct{}{}
+		flushed <- struct{}{}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return errors.Wrap(ctx.Err(), "long-running memtable flush in progress")
+	case <-flushed:
+		if b.active == nil && b.flushing == nil {
+			return nil
+		}
+
+		return b.FlushAndSwitch()
```

Comment:
What about `b.isReadOnly()` check? Inside flush goroutine it has to be false for flash to happen?

Reply:
I need to confirm with @etiennedi what the expected behavior is for a snapshot call when a class has readonly shards.